packages/@lwc/engine-core/src/framework/template.ts

```diff
@@ -110,6 +111,68 @@ function validateLightDomTemplate(template: Template, vm: VM) {
     }
 }
 
+// This function serializes the expressions generated by static content optimization.
+// Currently this is only needed for SSR.
+function buildSerializeExpressionFn(parts?: VStaticPart[]) {
+    if (process.env.IS_BROWSER || isUndefined(parts)) {
+        return () => '';
+    }
+    const partIdsToParts = new Map<string, VStaticPart>();
+    for (const staticPart of parts) {
+        partIdsToParts.set(`${staticPart.partId}`, staticPart);
+    }
+    return (partToken: string) => {
+        // The partTokens are split into 3 section:
+        // 1. The first character represents the expression type (attribute, class, style, or text).
+        // 2. The characters from index 1 to the first occurrence of a ':' is the partId.
+        // 3. Everything after the first ':' represents the attribute name.
+        // For example: a0:data-name, a = an attribute, 0 = partId, data-name = attribute name.
+        const type = partToken.charAt(0);
+        const delimiterIndex = partToken.indexOf(':');
+        const partId = partToken.substring(1, delimiterIndex);
+        const rest = partToken.substring(delimiterIndex + 1);
+        const part = partIdsToParts.get(partId) ?? EmptyObject;
+
+        let result = '';
+        switch (type) {
+            case 'a':
```

Comment:
I'm usually not a fan of enums in JavaScript, but I think this might be a good place to extract these letters (a, c, s, t) into enums in `@lwc/shared` so that it's easier to track where they come from.

Reply:
[Added](https://github.com/salesforce/lwc/pull/4056/commits/93f51d8c0d62013ab61fb6f499ce79329df14da1#diff-a3053a93c92783f74d97fe4270806f5c2275c9aae1273a4d5aba5671736577afR8)!